### PR TITLE
Add webrtc

### LINKED ISSUE #53: Add webrtc server
https://rustrepo.com/repo/centricular-webrtcsink-rust-network-programming

### PR DESCRIPTION
**Pull request by joaoantoniocardoso**
Fixes #53.

### CODE REVIEW

File: src/video_stream/types.rs [hunk 1]
```diff
@@ -38,7 +38,7 @@ impl VideoAndStreamInformation {
         let other_endpoints: HashSet<_> = other.stream_information.endpoints.iter().collect();
         let common_endpoints: HashSet<_> = our_endpoints.intersection(&other_endpoints).collect();
 
-        if !common_endpoints.is_empty() {
+        if !common_endpoints.is_empty() && self.stream_information.endpoints[0].scheme() != "turn" {
```
Comment (patrickelectric) [inline src/video_stream/types.rs:41]: We should add a `allow_same_endpoint` bool for webrtc stream type. and avoid this check here.

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: src/stream/turn_server.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        let mut turn_server = TURN_SERVER.as_ref().lock().unwrap();
+        turn_server.run = true;
+        turn_server.running_peers += 1;
+    }
+
+    pub fn notify_stop() {
+        debug!("TURN server was notified with a Stop!");
+        let mut turn_server = TURN_SERVER.as_ref().lock().unwrap();
+
+        if turn_server.running_peers != 0 {
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply src/stream/turn_server.rs:98]: it's unsigned.

File: src/stream/turn_server.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            turn_server.running_peers -= 1;
+        }
+
+        if turn_server.running_peers == 0 {
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply src/stream/turn_server.rs:102]: it's unsigned.

File: src/stream/manager.rs [hunk 4]
```diff
@@ -78,31 +81,41 @@ pub fn add_stream_and_start(
...
-        },
-    };
 
-    stream.mut_inner().start();
-    manager.streams.push(Stream {
-        stream_type: stream,
-        video_and_stream_information: video_and_stream_information.clone(),
-        mavlink_camera: MavlinkCameraHandle::new(
+    let mavlink_camera_handle = if let StreamType::WEBRTC(_) = &stream {
+        None
```
Comment (patrickelectric) [inline src/stream/manager.rs:88]: You can use `webrtc://<ice-user>:<ice-password>@<Address/Port>?<parameters>` for the video uri in mavlink manager.

File: src/stream/stream_backend.rs [hunk 5]
```diff
@@ -136,6 +137,26 @@ fn check_scheme(
                     return Err(SimpleError::new(format!("Endpoint with udp265 scheme only supports H265 encode. Encode: {:?}, Endpoints: {:#?}", encode, endpoints)));
                 }
             }
+            "turn" => {
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply src/stream/stream_backend.rs:141]: @joaoantoniocardoso what was the conclusion of that ?

[the remark this answers is not in the document]
Comment (patrickelectric) [reply src/stream/stream_backend.rs:141]: Oh, that the user could use turn or stun, right ?

File: src/stream/stream_backend.rs [hunk 6]
```diff
@@ -370,6 +387,109 @@ fn create_redirect_stream(
...
+    }
+
+    let configuration = match &video_and_stream_information
+        .stream_information
+        .configuration
+    {
+        CaptureConfiguration::VIDEO(configuration) => configuration,
+        _ => return Err(SimpleError::new(format!(
+            "The only accepted configuration type is the VideoCaptureConfiguration, which is like {supported_configuration} ",
+            supported_configuration="VideoCaptureConfiguration { \"encode\": <string>, \"height\": <integer>, \"width\": <integer>, \"frame_interval\": <integer/integer> }",
```
Comment (patrickelectric) [inline src/stream/stream_backend.rs:413]: Isn't a better way of doing it ?

File: src/stream/stream_backend.rs [hunk 7]
```diff
@@ -370,6 +387,109 @@ fn create_redirect_stream(
...
+            "The only accepted endpoint is: {expected_endpoint}, but {endpoint} was given."
+        )));
+    }
+
+    let configuration = match &video_and_stream_information
+        .stream_information
+        .configuration
+    {
+        CaptureConfiguration::VIDEO(configuration) => configuration,
+        _ => return Err(SimpleError::new(format!(
```
Comment (patrickelectric) [inline src/stream/stream_backend.rs:411]: Can you print what was received ?

File: src/stream/stream_backend.rs [hunk 8]
```diff
@@ -370,6 +387,109 @@ fn create_redirect_stream(
...
+
+    let encode = configuration.encode.clone();
+    let video_source = &video_and_stream_information.video_source;
+    let pipeline = match video_source {
+        VideoSourceType::Local(local_device) => {
+            if VideoEncodeType::H264 == encode {
+                format!(
+                    concat!(
+                        "v4l2src device={device}",
+                        " ! video/x-h264,width={width},height={height},framerate={interval_denominator}/{interval_numerator}",
```
Comment (patrickelectric) [inline src/stream/stream_backend.rs:425]: Create the caps in a previous line and use it here to avoid duplicated lines.

File: src/stream/stream_backend.rs [hunk 9]
```diff
@@ -386,6 +506,7 @@ fn create_stream(
         match endpoint.scheme() {
             "udp" => create_udp_stream(video_and_stream_information),
             "rtsp" => create_rtsp_stream(video_and_stream_information),
+            "turn" => create_webrtc_stream(video_and_stream_information),
```
Comment (patrickelectric) [inline src/stream/stream_backend.rs:510]: `create_webrtc_turn_stream`

File: src/stream/turn_server.rs [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+            Err(Error::ErrFakeErr)
+        }
+    }
+}
+
+lazy_static! {
+    pub static ref TURN_SERVER: Arc<Mutex<TurnServer>> =
+        Arc::new(Mutex::new(TurnServer::default()));
+}
+
```
Comment (patrickelectric) [inline src/stream/turn_server.rs:62]: We should have a public function to access the url of the turn server, with that we can avoid the hardcoded url around the source code for it.

File: src/stream/manager.rs [hunk 11]
```diff
@@ -136,6 +136,7 @@ pub fn remove_stream(stream_name: &str) -> Result<(), SimpleError> {
     let mut manager = MANAGER.as_ref().lock().unwrap();
     match manager.streams.iter().position(find_stream) {
         Some(index) => {
+            manager.streams[index].stream_type.mut_inner().stop();
```
Comment (patrickelectric) [inline src/stream/manager.rs:139]: That's a bit weird, why the stream can't stop itself on the desconstructor ?

File: src/server/manager.rs [hunk 12]
```diff
@@ -22,6 +23,8 @@ fn json_error_handler(error: JsonPayloadError, _: &HttpRequest) -> actix_web::Er
 
 // Start REST API server with the desired address
 pub fn run(server_address: &str) {
+    SignallingServer::start();
```
Comment (patrickelectric) [inline src/server/manager.rs:26]: Can you explain a bit better why it should start before the http server ?

**Review by patrickelectric — CHANGES_REQUESTED** [review]
The CI is also failing

File: src/stream/stream_backend.rs (hunk 6 above)
Comment (joaoantoniocardoso) [reply src/stream/stream_backend.rs:413]: Probably... but not for cheap. I'd asked about it in the BR rust community and it seems that the only solution to make this automatic would be to create a specific macro for that. I'd tried and it feels just too much effort for now... My approach was basically to create a macro that adds a custom implementation of a function that prints <name>: <type> of each field, but to use it, we would need to wrap all structs with that macro - that's ugly.

File: src/stream/manager.rs (hunk 11 above)
Comment (joaoantoniocardoso) [reply src/stream/manager.rs:139]: How would the stream know that it needs to be deconstructed once the client has clicked to 'remove' the stream? That's why I needed to add this here, and to me, it seems coherent with the `start()` called in the `add_stream_and_start()`.

File: src/server/manager.rs (hunk 12 above)
Comment (joaoantoniocardoso) [reply src/server/manager.rs:26]: I just thought that if the signaling server started _after_ the http, the frontend could request the signaling server before it's ready - which it's not a big deal, but can be avoided.

File: src/stream/stream_backend.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply src/stream/stream_backend.rs:141]: Supposedly yes, but I hadn't implemented that yet. The idea is that the user could pass an arbitrary stun and/or turn servers to be used instead of our local one. Now we have the signalling server too, which, in theory, could also be created by the user if it uses the [`protocol` specified here](https://github.com/centricular/webrtcsink/tree/main/protocol).

To implement this behavior we would need to use our local default unless it's specified by the user, and also identify as a webrtc stream for if a `stun://`, `turn://` or `ws://` is present in the list of endpoints. What do you think?

[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply src/stream/stream_backend.rs]: Okay, I did the changes, take a look now - the client can just use `webrtc://` (without host/ip) to use our local servers, or he can specify arbitrary turn, stun, and signalling servers to be used instead.

File: src/stream/manager.rs (hunk 4 above)
Comment (joaoantoniocardoso) [reply src/stream/manager.rs:88]: I didn't get the idea, I'd looked for any webrtc thing in mavlink and couldn't find anything, can you provide a reference of that?

**Comment by joaoantoniocardoso** [comment]
> The CI is also failing

yes, I'm not sure if we can publish our project using a dependency that is not published yet (webrtcsink). Perhaps we'd ask them about release plans..?

---

What was tested:
- WebRTC on Edge (Linux), Chrome (Linux, Android), and Firefox (Linux, Android).
- our default TURN/STUN/SIGNALLING servers.
- custom remote TURN/STUN servers.

What does not work:
- WebRTC without an interface - this is a browser limitation. I haven't tested it with a local network (between desktop and raspberry), but it should work.

What needs to be done:
- the client html/js. (I'd need help with the routing stuff in server/manager.rs)
- test with raspberry.

File: src/stream/manager.rs (hunk 11 above)
Comment (patrickelectric) [reply src/stream/manager.rs:139]: When you remove something from the list, the item will not have a owner and will call the constructor, there is no need to call stop manually, it should be done in the desconstructor.

File: src/server/manager.rs (hunk 12 above)
Comment (patrickelectric) [reply src/server/manager.rs:26]: So, can you add a comment about it ?

File: Cargo.toml [hunk 13]
```diff
@@ -50,11 +50,20 @@ pnet = "0.29.0"
...
+## GSTREAMER
 glib = { version = "0.15.2", optional = true }
 gstreamer = { version = "0.18.1", optional = true }
 gstreamer-rtsp-server = { version = "0.18.0", optional = true }
 
+## WebRTC Turn/Stun and Signalling server
+anyhow = "1"
+async-std = { version = "1", features = ["unstable", "attributes"] }
+async-native-tls = "0.3"
+tokio = { version = "1.17.0", features = ["full"] }
```
Comment (patrickelectric) [inline Cargo.toml:62]: just to be sure, do we need to have full features ?

File: src/stream/manager.rs (hunk 11 above)
Comment (joaoantoniocardoso) [reply src/stream/manager.rs:139]: Hmm, interesting design, I don't know why, but it is not how it is working. I'll try to debug it.

File: src/stream/stream_backend.rs [hunk 14]
```diff
@@ -136,6 +137,26 @@ fn check_scheme(
                     return Err(SimpleError::new(format!("Endpoint with udp265 scheme only supports H265 encode. Encode: {:?}, Endpoints: {:#?}", encode, endpoints)));
                 }
             }
+            "turn" => {
+                if VideoEncodeType::H264 != encode {
+                    return Err(SimpleError::new(format!("Endpoint with WebRTC scheme only supports H264 encode. Encode: {:?}, Endpoints: {:#?}", encode, endpoints)));
+                }
+
+                //WebRTC endpoints should contain user, credential, host and port
```
[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply src/stream/stream_backend.rs]: Now the front-end can simply pass `webrtc://` as the endpoint.

File: src/stream/stream_backend.rs [hunk 15]
```diff
@@ -370,6 +391,117 @@ fn create_redirect_stream(
...
+    video_and_stream_information: &VideoAndStreamInformation,
+) -> Result<StreamType, SimpleError> {
+    let endpoint = &video_and_stream_information.stream_information.endpoints[0];
+    if endpoint.scheme() != "turn" {
+        return Err(SimpleError::new(format!(
+            "The URL's scheme for WebRTC endpoints should be \"turn\", but was: {:?}",
+            endpoint.scheme()
+        )));
+    }
+    if endpoint.host_str() != "0.0.0.0".into() {
```
[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply src/stream/stream_backend.rs]: Thanks, the logic has changed entirely, so this is not relevant anymore.

File: src/stream/manager.rs (hunk 4 above)
Comment (patrickelectric) [reply src/stream/manager.rs:88]: You can add an url for the mavlink camera manager and not use none, having a mavlink camera handle for the webrtc stream.

File: src/stream/stream_backend.rs [hunk 16]
```diff
@@ -37,10 +41,14 @@ fn check_endpoints(
         return Err(SimpleError::new("Endpoints are empty".to_string()));
     }
 
+    let is_custom_webrtc = (endpoints[0].scheme() == "stun")
+        || (endpoints[0].scheme() == "turn")
+        || (endpoints[0].scheme() == "ws");
```
Comment (patrickelectric) [inline src/stream/stream_backend.rs:46]: ws stands for websocket and for not webrtc, is this correct ? should we use "webrtc" ?

File: src/stream/stream_backend.rs [hunk 17]
```diff
@@ -136,6 +144,23 @@ fn check_scheme(
                     return Err(SimpleError::new(format!("Endpoint with udp265 scheme only supports H265 encode. Encode: {:?}, Endpoints: {:#?}", encode, endpoints)));
                 }
             }
+            "webrtc" | "stun" | "turn" | "ws" => {
```
Comment (patrickelectric) [inline src/stream/stream_backend.rs:147]: Check my previous comment https://github.com/bluerobotics/mavlink-camera-manager/pull/60/files#r836481813

Comment (patrickelectric) [inline src/stream/stream_backend.rs:147]: We seriously need to handle that without the need of strings. We can do that in another PR.

File: src/stream/stream_backend.rs (hunk 6 above)
Comment (patrickelectric) [reply src/stream/stream_backend.rs:433]: It would be helpful that serde could provide that for us, but we can take a better look on the future.

**Comment by patrickelectric** [comment]
> Perhaps we should ask them about release plans..?

We should at least disable the deployment check step for now.

File: src/stream/stream_backend.rs (hunk 16 above)
Comment (patrickelectric) [reply src/stream/stream_backend.rs:46]: You can also do != webrtc

File: src/stream/manager.rs (hunk 11 above)
Comment (joaoantoniocardoso) [reply src/stream/manager.rs]: Perfect, I'd forgotten about implementing Drop() to call stop().

File: Cargo.toml (hunk 13 above)
Comment (joaoantoniocardoso) [reply Cargo.toml]: we needn't

File: Cargo.toml [hunk 18]
```diff
@@ -50,11 +50,20 @@ pnet = "0.29.0"
...
 
-## RTSP server
+## GSTREAMER
 glib = { version = "0.15.2", optional = true }
 gstreamer = { version = "0.18.1", optional = true }
 gstreamer-rtsp-server = { version = "0.18.0", optional = true }
 
+## WebRTC Turn/Stun and Signalling server
+anyhow = "1"
+async-std = { version = "1", features = ["unstable", "attributes"] }
```
[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply Cargo.toml]: you're right, thanks

File: src/stream/stream_backend.rs (hunk 17 above)
Comment (joaoantoniocardoso) [reply src/stream/stream_backend.rs:148]: indeed, and we already have plans to do that: https://github.com/bluerobotics/mavlink-camera-manager/issues/50

File: src/server/pages.rs [hunk 19]
```diff
@@ -64,22 +64,43 @@ pub fn load_file(file_name: &str) -> String {
     match file_name {
         "" | "index.html" => std::include_str!("../html/index.html").into(),
         "vue.js" => std::include_str!("../html/vue.js").into(),
+        "webrtc/index.html" => std::include_str!("../html/webrtc/index.html").into(),
```
Comment (patrickelectric) [inline src/server/pages.rs:67]: Be sure to remove all files that are incompatible with the MIT license, we can't include it at compile time.

Comment (patrickelectric) [inline src/server/pages.rs:67]: This function also needs to be updated to load files at runtime that is not the cargo folder, since we are going to download the webrtc files that are incompatible with our license.

File: src/stream/manager.rs (hunk 4 above)
Comment (joaoantoniocardoso) [reply src/stream/manager.rs]: Oh, I still need to do this part. Because webrtc definition is a combination of stun, turn, and signalling servers, how are we going to send it? I haven't checked yet, but we might only need the signalling server to be able to receive the stream somewhere else.

Comment (patrickelectric) [reply src/stream/manager.rs]: We could send only one of them, the first one provided.

**Review by patrickelectric — APPROVED** [review]
